New bullet system, PlasmaCutter

ReiderJack opened the pull request:
### Purpose
Bullets refactor, PlasmaCutter, hit animation for bullets, laser passing through windows.
fixes #4623
fixes #2307
fixes #2677

### Notes:
Proper laser sprite:
![Decals](https://user-images.githubusercontent.com/42844726/88978623-bcbcbf80-d2c8-11ea-8a9f-2b36fb8628c4.gif)

PlasmaCutter:
![Decals2](https://user-images.githubusercontent.com/42844726/88978627-bf1f1980-d2c8-11ea-9879-c1657efc399b.gif)

Impact from bullets:
![Decals3](https://user-images.githubusercontent.com/42844726/88978632-c21a0a00-d2c8-11ea-97c1-d1abc8e393ac.gif)

ReiderJack committed:
New bullet system, PlasmaCutter

fomalsd commented on UnityProject/Assets/Scripts/Weapons/PlasmaRefill.cs:
```diff
@@ -0,0 +1,43 @@
+		public void ServerPerformInteraction(HandApply interaction)
+		{
+			var needAmmo = magazineBehaviour.magazineSize - magazineBehaviour.ServerAmmoRemains;
+			if (needAmmo == 0) return;
```
```suggestion
			if (needAmmo <= 0) return;
```
perhaps add this just to be safe?

fomalsd commented on UnityProject/Assets/Scripts/Weapons/Projectiles/Beam.cs:
```diff
@@ -0,0 +1,52 @@
+				behaviour.OnShoot(direction, controlledByPlayer, fromWeapon, targetZone);
+			}
+
+			var pos = transform.position;
```
mm, transform.position seems a bit risky, do projectiles spawn at weapon's AssumedWorldPosServer?

ReiderJack replied:
The previous system used rigid body which was traveling on client side and server side. From what i tested it works fine. The Gun.cs spawns the bullet on client side by using `gameObject.transform.position`.

fomalsd replied:
ah, if this code is clientside then it’s all fine

fomalsd commented on UnityProject/Assets/Scripts/Weapons/Projectiles/Behaviours/BulletCasing.cs:
```diff
@@ -0,0 +1,23 @@
+			{
+				var netTransform = GetComponent<CustomNetTransform>();
+
+				netTransform?.ForceDrop(netTransform.ServerState.WorldPosition + new Vector3(Random.Range(-0.6f, 0.6f), Random.Range(-0.6f, 0.6f)));
```
```suggestion
if(netTransform != null)
{
netTransform.ForceDrop(netTransform.ServerState.WorldPosition + new Vector3(Random.Range(-0.6f, 0.6f), Random.Range(-0.6f, 0.6f)));
				}
```

ReiderJack replied:
Oh, this script was written by doobly, i just moved it to proper folder, lmao

ReiderJack replied:
Do you want me to commit this suggestion? I dunno if the change will work.

fomalsd replied:
this one absolutely won't hurt. formatting is f'ed up but blame github's text editor for that

fomalsd commented on UnityProject/Assets/Scripts/ScriptableObject/Gun/DespawnOnHit.cs:
```diff
@@ -0,0 +1,21 @@
+namespace Weapons.Projectiles.Behaviours
+{
+	/// <summary>
+	/// Will despawn object on any hit
```
where's the actual despawn code part? only see OnHit()

ReiderJack replied:
Sorry, this comment is misleading. This Script is supposed to return `bool` . So if it returns true, then it means that Hit processor wants the game object to be despawned.

fomalsd commented on UnityProject/Assets/Scripts/ScriptableObject/Gun/DespawnOnRequest.cs:
```diff
@@ -0,0 +1,25 @@
+namespace Weapons.Projectiles.Behaviours
+{
+	/// <summary>
+	/// Will despawn a game object only if a hit behaviour returns true
```
Same question here, maybe I don't understand something

fomalsd commented on UnityProject/Assets/Scripts/Weapons/Projectiles/Behaviours/ProjectileDamageAtPoint.cs:
```diff
@@ -0,0 +1,54 @@
+
+		private void RayCastAt(Vector2 point)
+		{
+			var rayHits = Physics2D.RaycastAll(point, direction,0f, layersToHit);
```
mmm, certainly wouldn't trust raycasts `for damaging everything on one tile`, you'd be better off using MatrixManager.GetAt<>(), or matrix.Get<> if you know a particular matrix

ReiderJack replied:
I was trying to start implementing logic for shooting like in old ss13. When you click on an object for example. If bullet travels to that point in the world without colliding with anything on its path, bullet will damage first object which is on this exact point where you clicked.

But i realized that i will have to touch firmly `Gun.cs` which i did not want to dig myself into.

fomalsd replied:
so what's your plan on this?

ReiderJack replied:
Leave it as it is. I will have to add another script with `MatrixManager.GetAt<>()` as you suggested. Or maybe your suggestion solves all problems and should be one script anyways.

The logic of clicking and shooting the point in world space where you clicked on, is supposed to get first object on the tile you clicked on.

fomalsd commented on UnityProject/Assets/Scripts/Weapons/Projectiles/Behaviours/ProjectileAnimatedDespawn.cs:
```diff
@@ -0,0 +1,82 @@
+
+		private IEnumerator EnableAnimationCoroutine(Vector3Int cellPos, TileChangeManager tileChangeManager, LayerTile oldEffectLayerTile)
+		{
+			tileChangeManager.UpdateTile(cellPos, TileType.Effects, animatedTile.name);
```
did you check if modifying cellPos z would mitigate the need to remove and reapply old effect?
I know matrices are 3d but don't remember if out network tile sync supports custom z values

ReiderJack replied:
I didn't check. Didn't dig this logic further tbh, just fixed the issues Proto-Kinetic-Accelerator had with this animation.

fomalsd commented on UnityProject/Assets/Scripts/Weapons/Projectiles/MovingProjectile.cs:
```diff
@@ -0,0 +1,80 @@
+
+			var hit = Raycast();
+
+			projectile.ProcessRaycastHit(hit);
```
why not inline `hit` variable?

ReiderJack replied:
You mean : `projectile.ProcessRaycastHit( Raycast());` ?
I think i did it cuz it was easy to see in debug what `var hit` is, but yeah.

ReiderJack replied:
You want to inline it?

fomalsd reviewed (commented):
I really like the behaviours approach, it's easy to understand yet very flexible.
There are just a few minor notes and questions.

I see a lot of GetComponent<> calls, I think shooting might be a bit expensive but that can be addressed later, having a solid maintainable system is more important. Note that I haven't playtested it yet.

Also curious if it's possible to make certain projectiles emit light

ReiderJack committed:
Remove redundant interface

IOnDespawn and IDespawn were too similar, merge them into one, works fine

ReiderJack committed:
IOnMove interface

Basically allows to have behaviors which are called when projectile moves.

ReiderJack committed:
Remove redundant check, fix formatting

ProjectileStun does need a check for shooter, cuz bullet already does that. Fixed BulletCasing formatting
